### PR TITLE
GEODE-6626: Add cache.gatewayreceiver.events.received counter

### PR DESCRIPTION
**Pull request by kirklund**
GEODE-6626: Add cache.gatewayreceiver.events.received counter

Inject GatewayReceiver and GatewayReceiverMetrics into ServerConnection
and intervening layers.

Increment GatewayReceiver eventsReceivedCounter in
GatewayReceiverCommand.

Co-authored-by: Mark Hanson <mhanson@pivotal.io>
Co-authored-by: Michael Oleske <moleske@pivotal.io>

There are also quite a few earlier commits which cleanup various classes and tests in preparation for adding the new counter.

Please review: @mhansonp @demery-pivotal @aaronlindsey

### CODE REVIEW

File: geode-core/src/integrationTest/java/org/apache/geode/internal/cache/tier/sockets/AcceptorImplIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+      acceptor1.close();
+    }
+    if (acceptor2 != null) {
+      acceptor2.close();
+    }
+    cache.close();
+  }
+
+  @Test
+  public void constructorThrowsBindException() throws CacheException, IOException {
```
Comment (demery-pivotal) [inline geode-core/src/integrationTest/java/org/apache/geode/internal/cache/tier/sockets/AcceptorImplIntegrationTest.java]: Under what conditions does the constructor throw? Add the condition(s) to the test name.

**Comment by kirklund** [comment]
GatewayReceiverMetrics is never closed. AcceptorImpl should close it when it closes GatewayReceiverStats.

File: geode-core/src/integrationTest/java/org/apache/geode/internal/cache/tier/sockets/AcceptorImplIntegrationTest.java (hunk 1 above)
Comment (kirklund) [reply geode-core/src/integrationTest/java/org/apache/geode/internal/cache/tier/sockets/AcceptorImplIntegrationTest.java]: I renamed the test to constructorThrowsBindExceptionIfPortIsInUse.

**Comment by kirklund** [comment]
Unfortunately @pivotal-jbarrett and our team both cleaned up and then modified ServerConnection so this conflict is a doozy.

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/command/GatewayReceiverCommand.java [hunk 2]
```diff
@@ -104,6 +96,8 @@ public void cmdExecute(final Message clientMessage, final ServerConnection serve
     int numberOfEvents = numberOfEventsPart.getInt();
     stats.incEventsReceived(numberOfEvents);
 
+    serverConnection.getGatewayReceiverMetrics().getEventsReceivedCounter().increment();
```
[the remark this answers is not in the document]
Comment (kirklund) [reply geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/command/GatewayReceiverCommand.java]: I believe you found an error. Thanks!

[the remark this answers is not in the document]
Comment (kirklund) [reply geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/command/GatewayReceiverCommand.java]: Fixed

**Comment by kirklund** [comment]
@upthewaterspout Unfortunately CacheServerImpl already contains GatewayReceiver logic in the form of the boolean so it was already tainted with poor design. I agree we should refactor this further. Eventually we should flip it around so that a GatewayReceiver (or InternalGatewayReceiver?) contains and uses a CacheServerImpl instead of passing it into CacheServerImpl. The GatewayReceiver would then create a GatewayReceiverAcceptor and pass that into CacheServerImpl as the Acceptor to be used. Mark and I were on the verge of tackling this but we felt that 2-3 more days of refactoring would be out of scope for one metric.

The discussion of stats vs metrics played out for weeks on our team. We definitely considered the approach that you're advocating but decided to go in the current direction. The mess known as CachePerfStats and RegionPerfStats provided much of the reason for this separation.

**Comment by kirklund** [comment]
@demery-pivotal I prefer to submit multiple PRs for cleaning up and refactoring code in addition to adding new functionality. Others prefer this approach. I'll push harder for the separate PRs next time.

**Comment by demery-pivotal** [comment]
> Maybe it would be better to encapsulate the micrometer updates within GatewayReceiverStats. Or better yet, maybe within the Statistics interface itself? We could enhance StatisticsFactory to have a way to indicate which stats get sent to which micrometer meter. Or consider looking at how our JMX mbeans did things - you don't see any JMX related code in GatewayReceiverCommand, but the MBeans are also getting this same events received data!

If the stat already measures exactly the information we want to publish through the meter, that's a fine solution. Another possibility is to use a style of meter that fetches its values from the stat (using a Gauge, FunctionCounter, or FunctionTimer).

Sometimes the stat doesn't measure exactly the thing we want to publish through the meters. For example (from an earlier meter PR), some of the put-related measure internal details that are meaningful to developers, but measure something other than what an operator would expect them to mean. In those cases, we would have to instrument the meter separately from the stats.

So… if the intentions of a stat and a meter are to measure _**exactly**_ the same thing, yeah, let's take advantage of that.

**Comment by demery-pivotal** [comment]
> @demery-pivotal I prefer to submit multiple PRs for cleaning up and refactoring code in addition to adding new functionality. Others prefer this approach. I'll push harder for the separate PRs next time.

As a developer, I very much prefer cleaning up as I go. As a reviewer, I might have a different preference sometimes. In this case, it's possible that a more comprehensive description would have sufficed for me, so that I would have a clearer idea of the intention of the changes in each file.

**Comment by kirklund** [comment]
@upthewaterspout I'll start a branch to separate CacheServerImpl and GatewayReceiver. Maybe we can follow up this PR with another that improves the design of CacheServerImpl.

**Comment by upthewaterspout** [comment]
> Sometimes the stat doesn't measure exactly the thing we want to publish through the meters. For example (from an earlier meter PR), some of the put-related measure internal details that are meaningful to developers, but measure something other than what an operator would expect them to mean. In those cases, we would have to instrument the meter separately from the stats.

It still makes more sense to have all of the stat updates go through a single consistent stat interface as far as the rest of the product is concerned. For the case of a stat that goes to micrometer but not VSD (why not just also send it to VSD, btw?) you could still add a method to GatewaySenderStats for that stat. Or StatisticsTypeFactory.createXXXCounter could in addition to taking a parameter indicating that something should go to micrometer, also indicate the stat doesn't need to go to VSD.

> So… if the intentions of a stat and a meter are to measure exactly the same thing, yeah, let's take advantage of that.

This should most of this micrometer work much easier and less error prone as well. Once you add some support in StatisticsFactory for micrometer, for any new stat that goes to micrometer you will just need to touch objects like GatewaySenderStats and add a flag or whatever during the statistics type creation indicating the stat should go to micrometer.

**Comment by moleske** [comment]
> (why not just also send it to VSD, btw?)

I believe that in general, when adding micrometer based metrics, we are not trying to change or enhance VSD based information.  The idea being that we will remove VSD once micrometer is providing all the metrics that VSD has.  This does not mean we couldn't add to VSD, but VSD is not tool of choice moving forward.

**Pull request closed without merging**